compiler/rustc_middle/src/ty/consts/kind.rs

```diff
@@ -26,6 +26,36 @@ impl<'tcx> UnevaluatedConst<'tcx> {
     pub fn expand(self) -> mir::UnevaluatedConst<'tcx> {
         mir::UnevaluatedConst { def: self.def, args: self.args, promoted: None }
     }
+
+    pub(crate) fn prepare_for_eval(
```

Comment:
I have absolutely no clue what the point of this code is, or why it is correct to *entirely disregard* the param_env when `has_non_region_infer` is true. This is just copied from the ty::const eval logic.

Should we do something similar for mir consts? Previously we only applied this for MIR consts that are ty consts, so that's what this PR sticks to.

Reply:
yea, no clue either, it's sufficiently scarily documented now so that we can look at it in the future. Will need to look at a few examples that actually make a difference here